### PR TITLE
sharding snapshot & journal cleanup

### PR DESCRIPTION
**Pull request by zbynek001**
port of
https://github.com/akka/akka/issues/21725
https://github.com/akka/akka/issues/22916

### CODE REVIEW

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs [hunk 1]
```diff
@@ -69,7 +69,46 @@ public class PersistentShard : Shard
...
+                    * for debugging if something goes wrong in production it's very useful to be able to inspect the events
+                    */
+                    var deleteToSequenceNr = m.Metadata.SequenceNr - Settings.TunningParameters.KeepNrOfBatches * Settings.TunningParameters.SnapshotAfter;
+                    if (deleteToSequenceNr > 0)
+                    {
+                        DeleteMessages(deleteToSequenceNr);
+                    }
+                    break;
+                case SaveSnapshotFailure m:
+                    Log.Warning("PersistentShard snapshot failure: {}", m.Cause.Message);
```
Comment (Horusiath) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs]: In .NET logging we must provide index for arguments. So `{}` itself will fail, you need to use `{0}`, `{1}` and so on.

Comment (zbynek001) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs]: Right, completely overlooked that. Fixed

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs [hunk 2]
```diff
@@ -69,7 +69,46 @@ public class PersistentShard : Shard
...
+                    break;
+
+                case DeleteMessagesFailure m:
+                    Log.Warning("PersistentShard messages to {0} deletion failure: {1}", m.ToSequenceNr, m.Cause.Message);
+                    break;
+                case DeleteSnapshotsSuccess m:
+                    Log.Debug("PersistentShard snapshots matching {0} deleted successfully", m.Criteria);
+                    break;
+                case DeleteSnapshotsFailure m:
+                    Log.Warning("PersistentShard snapshots matching {0} deletion falure: {1}", m.Criteria, m.Cause.Message);
```
Comment (alexvaluyskiy) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs]: Should be `failure`

**Comment by alexvaluyskiy** [comment]
@zbynek001 You have also implemented this issue
https://github.com/akka/akka/pull/20038
Because we did not handle snapshots before

**Comment by zbynek001** [comment]
I think that's obsolete code.
SaveSnapshotSuccess & SaveSnapshotFailure is already handled inside PersistentShard

File: src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs (hunk 2 above)
Comment (zbynek001) [inline src/contrib/cluster/Akka.Cluster.Sharding/PersistentShard.cs]: fixed

**Pull request merged**